### PR TITLE
Mop up liquids when placing furniture

### PR DESCRIPTION
**Pull request by JXMss**
#### Summary

Features "Mop up liquids when deploying furniture instead of interrupting the placement of furniture."

#### Purpose of change

The method of https://github.com/CleverRaven/Cataclysm-DDA/pull/53724 (checking for items on tiles when placing furniture) is simple and effective, but I feel that the detection is too broad and rough, and breaks some existing or future gameplay mechanics that will be implemented, such as simple item hiding (to avoid NPCs taking supplies at will), or using the planting system to simulate regular output actions such as beekeeping, composting or other no-agricultural actions(which would have to be directly masked by the deploy_furniture function to avoid letting the player get the seeds), etc.

Secondly, in reality, quite a lot of furniture is placed without damaging items on the floor (unless you just smash them down), such as tables, beds (which often have a lot of space near the floor), chairs, etc. However, liquids generally evaporate and disappear, so I chose to make them disappear in the game via the mop_spills() function, without destroying the original PR's target.

In addition, this change will reduce the amount of manipulation required when placing furniture, and we all know that the need for too much manipulation can be a hindrance to the player's play.

#### Describe the solution

~~Use the mop_spills() function instead of the has_items() function for the purpose of checking only liquids and ignoring other items.~~
Use tool_qualities to detect the presence of a mop within a certain range when placing furniture. If there is, the mop is used to remove the liquid from the floor; if not, the player is prompted that a mop is missing.
In order to avoid players stealing NPC items by using cardboard boxes and other furnishable homes, the determination of the ownership of items on the ground has been added. If there are items on the ground that do not belong to the player, it will stop the deployment of furniture and prompt the player.

#### Describe alternatives you've considered

Selective removal of liquid items using the item's liquid flag (a more complex implementation than the above method)

#### Testing

1. Change the code
2. Compiling CDDA with MSYS2
3. Separately place 8 liquid items and other items on the ground
4. Then use the "folded butter churn" to place the furniture on top of all the items separately
5. Use the map function in the debug menu to check if items on the ground have disappeared
6. Eventually all 8 liquid items were found to have disappeared and the other 8 other items were still there

### CODE REVIEW

File: src/iuse_actor.cpp [hunk 1]
```diff
@@ -1078,10 +1078,12 @@ cata::optional<int> deploy_furn_actor::use( Character &p, item &it, bool,
...
         return cata::nullopt;
     }
-    if( here.has_items( pnt ) ) {
-        p.add_msg_if_player( m_info, _( "Before deploying furniture, you need to clear the tile." ) );
-        return cata::nullopt;
+    
+    if( here.mop_spills( tripoint_bub_ms( pnt ) ) ) {
+        p.add_msg_if_player( m_info, _( "You moped up the spill when deploying furniture." ) );
+        p.moves -= 35;
     }
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp:1085]: You need to add check for a mop in player's possession. Otherwise liquids will be mopped by magic, without any mop.

Comment (JXMss) [reply src/iuse_actor.cpp:1085]: Can I avoid checking the mop if I change the message to "The liquid is slowly evaporating"? Or just clean up the water without generating a prompt at all.

Comment (JXMss) [reply src/iuse_actor.cpp:1085]: I was going to simulate the player using their hands or feet to clear the water away by increasing the time consumed by the action(The original time consumed for mopping was 15, and I changed it to 35).

Comment (JXMss) [reply src/iuse_actor.cpp:1085]: After all, in the real world, water evaporates and there is currently no such mechanism in the game.

Comment (JXMss) [reply src/iuse_actor.cpp]: I have now replaced the prompt message.

**Comment by PatrikLundell** [comment]
This would allow players to mop up liquids by placing furniture on top of it and then remove the furniture to magic the liquid away. A mop ought to be required to mop up liquids (you could, of course, introduce other means to mop up liquids as alternative methods, but placing furniture on top of a patch of sticky syrup and then remove the furniture will still leave a sticky patch in real life, even if it eventually gets drier).

**Comment by JXMss** [comment]
> This would allow players to mop up liquids by placing furniture on top of it and then remove the furniture to magic the liquid away. A mop ought to be required to mop up liquids (you could, of course, introduce other means to mop up liquids as alternative methods, but placing furniture on top of a patch of sticky syrup and then remove the furniture will still leave a sticky patch in real life, even if it eventually gets drier).

yes, you are right, I forgot about liquids like syrup. What a pity, it could have become so simple :(

**Comment by Night-Pryanik** [comment]
No need to add MOP qualities. MOP flag is already doing this.

**Comment by JXMss** [comment]
> No need to add MOP qualities. MOP flag is already doing this.

I want to use the "tool_qualities" to determine not only whether there is a mop on the player, but also whether there is a mop around.
And it has a complete range of functions ready to use(Mr. KorGgenT taught me that).

**Comment by Night-Pryanik** [comment]
Ok, I'm not against it, but anyway, simply adding qualities in json won't work because 
1) there's no corresponding code in source files related to this particular quality, i.e. right now `MOP` quality is just an ordinary text string in json file that does literally nothing.
2) there's still no check for this quality, i.e. even if `MOP` quality actually does anything, your changes doesn't check for its existence.

Also, I'd recommend you not to push your changes directly into your `master` branch as it will mess your git history, and create a separate branch for each new PR.

**Comment by JXMss** [comment]
> Ok, I'm not against it, but anyway, simply adding qualities in json won't work because
> 
> 1. there's no corresponding code in source files related to this particular quality, i.e. right now `MOP` quality is just an ordinary text string in json file that does literally nothing.
> 2. there's still no check for this quality, i.e. even if `MOP` quality actually does anything, your changes doesn't check for its existence.
> 
> Also, I'd recommend you not to push your changes directly into your `master` branch as it will mess your git history, and create a separate branch for each new PR.

1. I'm actually eating dinner right nowXD
2. Got it, thanks for pointing it out:)

File: src/iuse_actor.cpp (hunk 1 above)
Comment (JXMss) [reply src/iuse_actor.cpp]: Detection of mops has now been added, allowing you to mop up liquids when deploying furniture as long as the mop is within a certain radius of the player.

File: src/iuse_actor.cpp [hunk 2]
```diff
@@ -1078,10 +1079,18 @@ cata::optional<int> deploy_furn_actor::use( Character &p, item &it, bool,
         p.add_msg_if_player( m_info, _( "There is already furniture at that location." ) );
         return cata::nullopt;
     }
-    if( here.has_items( pnt ) ) {
-        p.add_msg_if_player( m_info, _( "Before deploying furniture, you need to clear the tile." ) );
-        return cata::nullopt;
+    
+    if( here.terrain_moppable( tripoint_bub_ms( pnt ) ) ) {
+        if( get_avatar().crafting_inventory().has_quality( qual_MOP ) ){
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp:1084]:
```suggestion
        if( get_avatar().crafting_inventory().has_quality( qual_MOP ) ) {
```

Comment (JXMss) [reply src/iuse_actor.cpp]: Thx! I didn't notice it at allXD

**Comment by JXMss** [comment]
Now it just needs to check the Validate PR, could you help me get through the rerun request?:P @Night-Pryanik

**Comment by andrei8l** [comment]
This lets you easily steal items from NPCs:
1. Deploy a Large Cardboard over the NPC's items. Break the existing furniture if needed
2. Items are now magically inside the box
3. Drag box out of sight, along with its contained items
4. Steal to your heart's content

Can be tested easily in the Forge of Wonders or the Exodii base

> or using the planting system to simulate regular output actions such as beekeeping, composting or other no-agricultural actions(which would have to be directly masked by the deploy_furniture function to avoid letting the player get the seeds), etc.

These problems are better solved with constructions or update/nested mapgen and sealed furniture.

**Comment by JXMss** [comment]
> This lets you easily steal items from NPCs:
> 
> 1. Deploy a Large Cardboard over the NPC's items. Break the existing furniture if needed
> 2. Items are now magically inside the box
> 3. Drag box out of sight, along with its contained items
> 4. Steal to your heart's content
> 
> Can be tested easily in the Forge of Wonders or the Exodii base

I think the theft of items should be addressed by the NPC system. For example, removing and destroying furniture belonging to those fixed NPCs should be equivalent to players stealing items. 
Also, it's much easier for players to move items this way and reduces a lot of manipulation. Most players don't want to spend a lot of real time doing something as tedious as hauling a lot of items (most CDDA players I know feel this way, especially mobile players).


> 
> > or using the planting system to simulate regular output actions such as beekeeping, composting or other no-agricultural actions(which would have to be directly masked by the deploy_furniture function to avoid letting the player get the seeds), etc.
> 
> These problems are better solved with constructions or update/nested mapgen and sealed furniture.

I couldn't agree more with this, but I'm just a C++ novice myself and don't have the skills to tackle this for now:(

**Comment by JXMss** [comment]
> This lets you easily steal items from NPCs:
> 
> 1. Deploy a Large Cardboard over the NPC's items. Break the existing furniture if needed
> 2. Items are now magically inside the box
> 3. Drag box out of sight, along with its contained items
> 4. Steal to your heart's content
> 
> Can be tested easily in the Forge of Wonders or the Exodii base
> 
> > or using the planting system to simulate regular output actions such as beekeeping, composting or other no-agricultural actions(which would have to be directly masked by the deploy_furniture function to avoid letting the player get the seeds), etc.
> 
> These problems are better solved with constructions or update/nested mapgen and sealed furniture.

I have now added the detection of other people's items and after testing the function works well.

**Comment by JXMss** [comment]
Now that all the checks have been passed, do I just have to wait for a member with sufficient permissions to merge?

**Comment by Night-Pryanik** [comment]
Yes, exactly.

**Comment by JXMss** [comment]
> Yes, exactly.

Got it.

**Comment by JXMss** [comment]
sry to bother you, but could you help me rerun the PR request? much appreciated:). @Night-Pryanik

**Comment by Night-Pryanik** [comment]
Done.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/a-frustrating-update/28289/5

**Comment by JXMss** [comment]
Alas, the waiting is really painful:(

**Pull request merged**